qiskit_experiments/characterization/t1_experiment.py

```diff
@@ -0,0 +1,206 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+T1 Experiment class.
+"""
+
+from typing import List, Optional, Union, Tuple
+import numpy as np
+
+from qiskit.circuit import QuantumCircuit
+
+from qiskit_experiments.base_experiment import BaseExperiment
+from qiskit_experiments.base_analysis import BaseAnalysis
+from qiskit_experiments.analysis.curve_fitting import process_curve_data, curve_fit
+from qiskit_experiments.analysis.data_processing import level2_probability
+from qiskit_experiments import AnalysisResult
+
+
+class T1Analysis(BaseAnalysis):
+    """T1 Experiment result analysis class."""
+
+    # pylint: disable=arguments-differ, unused-argument
+    def _run_analysis(
+        self,
+        experiment_data,
+        t1_guess=None,
+        amplitude_guess=None,
+        offset_guess=None,
+        t1_bounds=None,
+        amplitude_bounds=None,
+        offset_bounds=None,
+        **kwargs,
+    ) -> Tuple[AnalysisResult, None]:
+        """
+        Calculate T1
+
+        Args:
+            experiment_data (ExperimentData): the experiment data to analyze
+            t1_guess (float): Optional, an initial guess of T1
+            amplitude_guess (float): Optional, an initial guess of the coefficient of the exponent
+            offset_guess (float): Optional, an initial guess of the offset
+            t1_bounds (list of two floats): Optional, lower bound and upper bound to T1
+            amplitude_bounds (list of two floats): Optional, lower bound and upper bound to the amplitude
+            offset_bounds (list of two floats): Optional, lower bound and upper bound to the offset
+            kwargs: Trailing unused function parameters
+
+        Returns:
+            The analysis result with the estimated T1
+        """
+
+        circuit_unit = experiment_data._data[0]["metadata"]["unit"]
+        dt_factor_in_sec = experiment_data._data[0]["metadata"].get("dt_factor_in_sec", None)
+        if dt_factor_in_sec is None:
+            dt_factor_in_microsec = 1
+            result_unit = circuit_unit
+        else:
+            dt_factor_in_microsec = dt_factor_in_sec * 1000000
+            result_unit = "us"
+
+        xdata, ydata, sigma = process_curve_data(
+            experiment_data._data, lambda datum: level2_probability(datum, "1")
+        )
```

Comment:
A couple of comments:
1. The old code checked for each sigma if it's equal to 0, and if yes, it assigned a small value to it. Perhaps `process_curve_data` or `level2_probability` should set very small sigmas to a larger (but still tiny) value.
2. Allow a flexible data processor here (I suggest in a follow-up PR, is it important to do it now?).